src/wayback/_http.py

```diff
@@ -0,0 +1,282 @@
+#
+# More info in this issue:
+# https://github.com/edgi-govdata-archiving/web-monitoring-processing/issues/309
+#
+# Example broken Wayback URL:
+# http://web.archive.org/web/20181023233237id_/http://cwcgom.aoml.noaa.gov/erddap/griddap/miamiacidification.graph
+#
+if hasattr(HTTPConnectionPool, 'ResponseCls'):
+    # urllib3 v1.x:
+    #
+    # This subclass of urllib3's response class identifies the malformed headers
+    # and repairs them before instantiating the actual response object, so when
+    # it reads the body, it knows to decode it correctly.
+    #
+    # See what we're overriding from urllib3:
+    # https://github.com/urllib3/urllib3/blob/a6ec68a5c5c5743c59fe5c62c635c929586c429b/src/urllib3/response.py#L499-L526
+    class WaybackUrllib3Response(HTTPConnectionPool.ResponseCls):
+        @classmethod
+        def from_httplib(cls, httplib_response, **response_kwargs):
+            headers = httplib_response.msg
+            pairs = headers.items()
+            if ('content-encoding', '') in pairs and ('Content-Encoding', 'gzip') in pairs:
+                del headers['content-encoding']
+                headers['Content-Encoding'] = 'gzip'
+            return super().from_httplib(httplib_response, **response_kwargs)
+
+    HTTPConnectionPool.ResponseCls = WaybackUrllib3Response
+else:
+    # urllib3 v2.x:
+    #
+    # Unfortunately, we can't wrap the `HTTPConnection.getresponse` method in
+    # urllib3 v2, since it may have read the response body before returning. So
+    # we patch the HTTPHeaderDict class instead.
+    from urllib3._collections import HTTPHeaderDict as Urllib3HTTPHeaderDict
+    _urllib3_header_init = Urllib3HTTPHeaderDict.__init__
+
+    def _new_header_init(self, headers=None, **kwargs):
+        if headers:
+            if isinstance(headers, (Urllib3HTTPHeaderDict, dict)):
+                pairs = list(headers.items())
+            else:
+                pairs = list(headers)
+            if (
+                ('content-encoding', '') in pairs and
+                ('Content-Encoding', 'gzip') in pairs
+            ):
+                headers = [pair for pair in pairs
+                           if pair[0].lower() != 'content-encoding']
+                headers.append(('Content-Encoding', 'gzip'))
+
+        return _urllib3_header_init(self, headers, **kwargs)
+
+    Urllib3HTTPHeaderDict.__init__ = _new_header_init
+# END HACK
+#####################################################################
+
+
+class WaybackHttpResponse:
+    """
+    Represents an HTTP response from a server. This might be included as an
+    attribute of an exception, but should otherwise not be exposed to user
+    code in normal circumstances. It's meant to wrap to provide a standard,
+    thread-safe interface to response objects from whatever underlying HTTP
+    tool is being used (e.g. requests, httpx, etc.).
+    """
+    status_code: int
+    headers: dict
+    encoding: Optional[str] = None
+    url: str
+    links: dict
```

Comment:
Wondering if this class should move to `_models`, especially if we make customizing `WaybackHttpAdapter` a thing. (Since you’ll need to make your own subclass of this to return from an adapter that doesn’t use *Requests*.)